lib/Autoload/UniversalClassLoader.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+class UniversalClassLoader {
+	protected $registeredPath = [];
+
+	/**
+	 * Register path in which to look for a class.
+	 */
+	public function registerPath($path) {
+		if (is_string($path)) {
+			$this->registeredPath[] = $path;
+		}
+		if (is_array($path)) {
+			array_push($this->registeredPath, ...$path);
+		}
+	}
+
+	/**
+	 * Indicates if the processed class is supported by the current loader
+	 */
+	protected function isSupported($class) {
+		return true;
+	}
+
+	/**
+	 * Load class file if found.
+	 */
+	public function loadClass($class) {
+		if (!$this->isSupported($class)) {
+			return;
+		}
+
+		if ($file = $this->findFile($class)) {
+			require_once $file;
+		}
+	}
+
+	/**
+	 * Find the file containing the class definition.
+	 */
+	protected function findFile($class) {
+		foreach ($this->registeredPath as $path) {
```

Comment:
I am still not a fan of searching on disk, something very slow (especially since it needs to be done for every PHP file we load), when we already know where it is...

Reply:
I can add caching on that as well.

Reply:
P.S. I know you have tried to explain higher up the rationale, but I still don't get it. It is significantly longer code (harder to understand, at least to me) with loops and disk scans, and it does not look like it solves any current problem, does it?